File: go/arrow/array/boolean_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package array_test

import (
	"reflect"
	"testing"

	"github.com/apache/arrow/go/arrow/array"
	"github.com/apache/arrow/go/arrow/memory"
)

func TestBooleanSliceData(t *testing.T) {
	pool := memory.NewCheckedAllocator(memory.NewGoAllocator())
	defer pool.AssertSize(t, 0)

	values := []bool{true, false, true, true, true, true, true, false, true, false}

	b := array.NewBooleanBuilder(pool)
	defer b.Release()

	for _, v := range values {
		b.Append(v)
	}

	arr := b.NewArray().(*array.Boolean)
	defer arr.Release()

	if got, want := arr.Len(), len(values); got != want {
		t.Fatalf("got=%d, want=%d", got, want)
	}

	vs := make([]bool, arr.Len())

	for i := range vs {
		vs[i] = arr.Value(i)
	}

	if got, want := vs, values; !reflect.DeepEqual(got, want) {
		t.Fatalf("got=%v, want=%v", got, want)
	}

	tests := []struct {
		interval [2]int64
		want     []bool
	}{
		{
			interval: [2]int64{0, 0},
			want:     []bool{},
		},
		{
			interval: [2]int64{10, 10},
			want:     []bool{},
		},
		{
			interval: [2]int64{0, 5},
			want:     []bool{true, false, true, true, true},
		},
		{
			interval: [2]int64{5, 10},
			want:     []bool{true, true, false, true, false},
		},
		{
			interval: [2]int64{2, 7},
			want:     []bool{true, true, true, true, true},
		},
	}

	for _, tc := range tests {
		t.Run("", func(t *testing.T) {

			slice := array.NewSlice(arr, tc.interval[0], tc.interval[1]).(*array.Boolean)
			defer slice.Release()

			if got, want := slice.Len(), len(tc.want); got != want {
				t.Fatalf("got=%d, want=%d", got, want)
			}

			vs := make([]bool, slice.Len())

			for i := range vs {
				vs[i] = slice.Value(i)
			}

			if got, want := vs, tc.want; !reflect.DeepEqual(got, want) {
				t.Fatalf("got=%v, want=%v", got, want)
			}
		})
	}
}

func TestBooleanSliceDataWithNull(t *testing.T) {
	pool := memory.NewCheckedAllocator(memory.NewGoAllocator())
	defer pool.AssertSize(t, 0)

	values := []bool{true, false, true, false, false, false, true, false, true, false}
	valids := []bool{true, false, true, true, true, true, true, false, true, true}

	b := array.NewBooleanBuilder(pool)
	defer b.Release()

	b.AppendValues(values, valids)

	arr := b.NewArray().(*array.Boolean)
	defer arr.Release()

	if got, want := arr.Len(), len(valids); got != want {
		t.Fatalf("got=%d, want=%d", got, want)
	}

	if got, want := arr.NullN(), 2; got != want {
		t.Fatalf("got=%d, want=%d", got, want)
	}

	vs := make([]bool, arr.Len())

	for i := range vs {
		vs[i] = arr.Value(i)
	}

	if got, want := vs, values; !reflect.DeepEqual(got, want) {
		t.Fatalf("got=%v, want=%v", got, want)
	}

	tests := []struct {
		interval [2]int64
		nulls    int
		want     []bool
	}{
		{
			interval: [2]int64{2, 9},
			nulls:    1,
			want:     []bool{true, false, false, false, true, false, true},
		},
		{
			interval: [2]int64{0, 7},
			nulls:    1,
			want:     []bool{true, false, true, false, false, false, true},
		},
		{
			interval: [2]int64{1, 8},
			nulls:    2,
			want:     []bool{false, true, false, false, false, true, false},
		},
		{
			interval: [2]int64{2, 7},
			nulls:    0,
			want:     []bool{true, false, false, false, true},
		},
	}

	for _, tc := range tests {
		t.Run("", func(t *testing.T) {

			slice := array.NewSlice(arr, tc.interval[0], tc.interval[1]).(*array.Boolean)
			defer slice.Release()

			if got, want := slice.NullN(), tc.nulls; got != want {
				t.Errorf("got=%d, want=%d", got, want)
			}

			if got, want := slice.Len(), len(tc.want); got != want {
				t.Fatalf("got=%d, want=%d", got, want)
			}

			vs := make([]bool, slice.Len())

			for i := range vs {
				vs[i] = slice.Value(i)
			}

			if got, want := vs, tc.want; !reflect.DeepEqual(got, want) {
				t.Fatalf("got=%v, want=%v", got, want)
			}
		})
	}
}

func TestBooleanSliceOutOfBounds(t *testing.T) {
	pool := memory.NewCheckedAllocator(memory.NewGoAllocator())
	defer pool.AssertSize(t, 0)

	values := []bool{true, false, true, false, true, false, true, false, true, false}

	b := array.NewBooleanBuilder(pool)
	defer b.Release()

	for _, v := range values {
		b.Append(v)
	}

	arr := b.NewArray().(*array.Boolean)
	defer arr.Release()

	slice := array.NewSlice(arr, 3, 8).(*array.Boolean)
	defer slice.Release()

	tests := []struct {
		index int
		panic bool
	}{
		{
			index: -1,
			panic: true,
		},
		{
			index: 5,
			panic: true,
		},
		{
			index: 0,
			panic: false,
		},
		{
			index: 4,
			panic: false,
		},
	}

	for _, tc := range tests {
		t.Run("", func(t *testing.T) {

			var val bool

			if tc.panic {
				defer func() {
					e := recover()
					if e == nil {
						t.Fatalf("this should have panicked, but did not; slice value %v", val)
					}
					if got, want := e.(string), "arrow/array: index out of range"; got != want {
						t.Fatalf("invalid error. got=%q, want=%q", got, want)
					}
				}()
			} else {
				defer func() {
					if e := recover(); e != nil {
						t.Fatalf("unexpected panic: %v", e)
					}
				}()
			}

			val = slice.Value(tc.index)
		})
	}
}

File: cpp/src/arrow/compute/kernels/boolean-test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <cstdint>
#include <functional>
#include <memory>
#include <vector>

#include <gtest/gtest.h>

#include "arrow/testing/gtest_common.h"
#include "arrow/testing/gtest_util.h"

#include "arrow/compute/context.h"
#include "arrow/compute/kernel.h"
#include "arrow/compute/kernels/boolean.h"
#include "arrow/compute/kernels/util-internal.h"
#include "arrow/compute/test-util.h"

using std::shared_ptr;
using std::vector;

namespace arrow {
namespace compute {

using BinaryKernelFunc =
    std::function<Status(FunctionContext*, const Datum&, const Datum&, Datum* out)>;

class TestBooleanKernel : public ComputeFixture, public TestBase {
 public:
  void TestArrayBinary(const BinaryKernelFunc& kernel, const std::shared_ptr<Array>& left,
                       const std::shared_ptr<Array>& right,
                       const std::shared_ptr<Array>& expected) {
    Datum result;
    ASSERT_OK(kernel(&this->ctx_, left, right, &result));
    ASSERT_EQ(Datum::ARRAY, result.kind());
    std::shared_ptr<Array> result_array = result.make_array();
    ASSERT_ARRAYS_EQUAL(*expected, *result_array);
  }

  void TestChunkedArrayBinary(const BinaryKernelFunc& kernel,
                              const std::shared_ptr<ChunkedArray>& left,
                              const std::shared_ptr<ChunkedArray>& right,
                              const std::shared_ptr<ChunkedArray>& expected) {
    Datum result;
    std::shared_ptr<Array> result_array;
    ASSERT_OK(kernel(&this->ctx_, left, right, &result));
    ASSERT_EQ(Datum::CHUNKED_ARRAY, result.kind());
    std::shared_ptr<ChunkedArray> result_ca = result.chunked_array();
    ASSERT_TRUE(result_ca->Equals(expected));
  }

  void TestBinaryKernel(const BinaryKernelFunc& kernel, const std::vector<bool>& values1,
                        const std::vector<bool>& values2,
                        const std::vector<bool>& values3,
                        const std::vector<bool>& values3_nulls) {
    auto type = boolean();
    auto a1 = _MakeArray<BooleanType, bool>(type, values1, {});
    auto a2 = _MakeArray<BooleanType, bool>(type, values2, {});
    auto a3 = _MakeArray<BooleanType, bool>(type, values3, {});
    auto a1_nulls = _MakeArray<BooleanType, bool>(type, values1, values1);
    auto a2_nulls = _MakeArray<BooleanType, bool>(type, values2, values2);
    auto a3_nulls = _MakeArray<BooleanType, bool>(type, values3, values3_nulls);

    TestArrayBinary(kernel, a1, a2, a3);
    TestArrayBinary(kernel, a1_nulls, a2_nulls, a3_nulls);
    TestArrayBinary(kernel, a1->Slice(1), a2->Slice(1), a3->Slice(1));
    TestArrayBinary(kernel, a1_nulls->Slice(1), a2_nulls->Slice(1), a3_nulls->Slice(1));

    // ChunkedArray
    std::vector<std::shared_ptr<Array>> ca1_arrs = {a1, a1->Slice(1)};
    auto ca1 = std::make_shared<ChunkedArray>(ca1_arrs);
    std::vector<std::shared_ptr<Array>> ca2_arrs = {a2, a2->Slice(1)};
    auto ca2 = std::make_shared<ChunkedArray>(ca2_arrs);
    std::vector<std::shared_ptr<Array>> ca3_arrs = {a3, a3->Slice(1)};
    auto ca3 = std::make_shared<ChunkedArray>(ca3_arrs);
    TestChunkedArrayBinary(kernel, ca1, ca2, ca3);

    // ChunkedArray with different chunks
    std::vector<std::shared_ptr<Array>> ca4_arrs = {a1->Slice(0, 1), a1->Slice(1),
                                                    a1->Slice(1, 1), a1->Slice(2)};
    auto ca4 = std::make_shared<ChunkedArray>(ca4_arrs);
    TestChunkedArrayBinary(kernel, ca4, ca2, ca3);
  }
};

TEST_F(TestBooleanKernel, Invert) {
  vector<bool> values1 = {true, false, true, false};
  vector<bool> values2 = {false, true, false, true};

  auto type = boolean();
  auto a1 = _MakeArray<BooleanType, bool>(type, values1, {true, true, true, false});
  auto a2 = _MakeArray<BooleanType, bool>(type, values2, {true, true, true, false});

  // Plain array
  Datum result;
  ASSERT_OK(Invert(&this->ctx_, a1, &result));
  ASSERT_EQ(Datum::ARRAY, result.kind());
  ASSERT_ARRAYS_EQUAL(*a2, *result.make_array());

  // Array with offset
  ASSERT_OK(Invert(&this->ctx_, a1->Slice(1), &result));
  ASSERT_EQ(Datum::ARRAY, result.kind());
  ASSERT_ARRAYS_EQUAL(*a2->Slice(1), *result.make_array());

  // ChunkedArray
  std::vector<std::shared_ptr<Array>> ca1_arrs = {a1, a1->Slice(1)};
  auto ca1 = std::make_shared<ChunkedArray>(ca1_arrs);
  std::vector<std::shared_ptr<Array>> ca2_arrs = {a2, a2->Slice(1)};
  auto ca2 = std::make_shared<ChunkedArray>(ca2_arrs);
  ASSERT_OK(Invert(&this->ctx_, ca1, &result));
  ASSERT_EQ(Datum::CHUNKED_ARRAY, result.kind());
  std::shared_ptr<ChunkedArray> result_ca = result.chunked_array();
  ASSERT_ARRAYS_EQUAL(*ca2, *result_ca);
}

TEST_F(TestBooleanKernel, InvertEmptyArray) {
  std::vector<std::shared_ptr<Buffer>> data_buffers(2);
  Datum input;
  input.value = ArrayData::Make(boolean(), 0 /* length */, std::move(data_buffers),
                                0 /* null_count */);

  Datum result;
  ASSERT_OK(Invert(&this->ctx_, input, &result));
  ASSERT_ARRAYS_EQUAL(*input.make_array(), *result.make_array());
}

TEST_F(TestBooleanKernel, BinaryOpOnEmptyArray) {
  auto type = boolean();
  std::vector<std::shared_ptr<Buffer>> data_buffers(2);
  Datum input;
  input.value = ArrayData::Make(boolean(), 0 /* length */, std::move(data_buffers),
                                0 /* null_count */);

  Datum result;
  ASSERT_OK(And(&this->ctx_, input, input, &result));
  // Result should be empty as well.
  ASSERT_ARRAYS_EQUAL(*input.make_array(), *result.make_array());
}

TEST_F(TestBooleanKernel, And) {
  vector<bool> values1 = {true, false, true, false, true, true};
  vector<bool> values2 = {true, true, false, false, true, false};
  vector<bool> values3 = {true, false, false, false, true, false};
  TestBinaryKernel(And, values1, values2, values3, values3);
}

TEST_F(TestBooleanKernel, Or) {
  vector<bool> values1 = {true, false, true, false, true, true};
  vector<bool> values2 = {true, true, false, false, true, false};
  vector<bool> values3 = {true, true, true, false, true, true};
  vector<bool> values3_nulls = {true, false, false, false, true, false};
  TestBinaryKernel(Or, values1, values2, values3, values3_nulls);
}

TEST_F(TestBooleanKernel, Xor) {
  vector<bool> values1 = {true, false, true, false, true, true};
  vector<bool> values2 = {true, true, false, false, true, false};
  vector<bool> values3 = {false, true, true, false, false, true};
  vector<bool> values3_nulls = {true, false, false, false, true, false};
  TestBinaryKernel(Xor, values1, values2, values3, values3_nulls);
}

}  // namespace compute
}  // namespace arrow
